### ISSUE TITLE
"topology.conf" leads to high CPU usage and memory not being released

### ISSUE THREAD

**Issue by luxiaoyong** [cea-hpc/clustershell#527]
My 'topology.conf' config file has ? or * in gateway names like below:

[routes]
Controller-vm1:Controller-vm2,Controller-vm3
Controller-vm2**?**:Computer101

After executing command: "env/bin/clush -d -w 127.1 whoami", an error occurred and the command is stuck, which leads to high CPU usage and memory not being released. The debug output is below:
```
DEBUG:root:clush: STARTING DEBUG
Changing max open files soft limit from 65535 to 8192
User interaction: True
Create STDIN worker: False
DEBUG:ClusterShell.NodeUtils:EXEC '[ -f /etc/clustershell/groups ] && f=/etc/clustershell/groups || f=/etc/clustershell/groups.d/local.cfg; sed -n 's/^all:\(.*\)/\1/p' $f'
DEBUG:ClusterShell.NodeUtils:READ ''
```

Please help me check this issue, thank you very much.

**Comment by degremont** [comment]
Hi @luxiaoyong 

Even if Clustershell should protect itself against such syntax and simply report an error, please note that this syntax is not supported.

Aurélien

**Comment by luxiaoyong** [comment]
I want to find out the reason why the gateway name with special characters will cause high cpu usage and will not be released.

**Comment by luxiaoyong** [comment]
Hi @degremont 
    Thank you very much for your suggestion.
    I read the source code yesterday and print messages in my environment, I found that if the ? in my gateway name, then curr.children has a empty string child in the code, and the empty child has two childs which include an empty string and another child, code endless loop caused by this reason. This is why the cpu usage surges.
    And this problem can be fixed with the following code, the code in Topology.py file:
`def load(self, rootnode):
    self.root = rootnode

    stack = [rootnode]

    while len(stack) > 0:
        curr = stack.pop()
        self.groups.append(curr)

        if curr.children_len() > 0:
            childs = curr.children()
            for i in range(len(childs) - 1, -1, -1):
                if not childs[i].nodeset:
                    del childs[i]

            stack += childs`
I hope that the above vulnerabilities can be fixed in the next version. I would appreciate it.

**Comment by degremont** [comment]
We had identified the bug in that area of the code but we did not had time to deep dive into it yet. Thanks for your help and analysis. We will look at this shortly and it is on our list for the next release.

Out of curiosity, there are multiple ways to avoid that problem, with different syntax in `topology.conf`. Why do espcially want this one here?

**Comment by luxiaoyong** [comment]
Hi @degremont
I am glad to receive your reply. The reason why we found this problem is that our topology.conf file is automatically generated. Special characters are accidentally added. After testing, we found it ? *&  three special characters cause the above problem. So far, we do not know whether there are other special characters that cause the problem. We can think of a way to avoid the problem by checking these special characters, but we need to know the full set of the special characters that cause the problem. Therefore, we can use the above code to delete the empty sub-node to avoid this problem. I look forward to seeing your optimization of this issue in the next release.
By the way, is there a way to verify the legality of the topology.conf? Thanks very much.
